### ISSUE TITLE
How to pass softmax activation output from RNN to custom optimizer

### ISSUE THREAD

**Issue by JustinS6626** [benoitdescamps/benoit-descamps-blogs#1]
I apologize if the topic of this post is not relevant to the blog, but I was not sure how else to contact use (I don't use Linkdin), and I was hoping to ask you a question about writing a custom Tensorflow optimizer, as it looks like you wrote an article on that subject fairly recently. The question that I have is one that I have tried posting to StackOverflow and on the Tensorflow discussion and developer forums, but have received no reply.

 I have written a three-layer gated recurrent unit RNN in Tensorflow, and I am trying to train it with a custom optimizer class that I am writing myself. This optimizer is meant to use the softmax output from the RNN as one of the parameters in the _apply_dense method of the optimizer class.

The softmax result is stored in the self.dist attribute of the RNN class, defined by

self.meas = tf.compat.v1.placeholder(tf.float32, [None, None, self.ndigits])

self.molecules, _ = self.model(scopeName=self.model_code+"_Init", taskName=self.task_code+"_Init")
self.dist, self.logP = self.model()
self.generation_loss = -tf.reduce_sum(self.molecules * tf.math.log(1e-10 + self.dist))

The self.model() function is defined as follows:

def model(self, scopeName=None, taskName=None):

    if scopeName == None:
        scope = self.model_code
    else:
        scope = scopeName

    if taskName == None:
        task = self.task_code
    else:
        task = taskName
    logP = tf.zeros([self.batchsize], dtype=tf.float32)

    with tf.compat.v1.variable_scope(scope):

        zt = tf.compat.v1.get_variable(task, [1, self.repsize])

        z = tf.tile(zt, [self.batchsize, 1])

        z_mat = tf.layers.dense(inputs=z, units=self.repsize, activation=tf.nn.relu, name="wMat"+task)

        h1 = tf.stack([z_mat for _ in range(self.nqubits)], axis=1)

        cell = tf.contrib.rnn.MultiRNNCell([tf.contrib.rnn.GRUCell(self.row_units) for _ in range(self.hidden_rows)], state_is_tuple=False)

        outputs, state = tf.nn.dynamic_rnn(cell=cell, inputs=h1, dtype=tf.float32)
        h2 = tf.layers.dense(tf.reshape(outputs, [self.batchsize*self.nqubits, self.row_units]), 
                             units=self.statevec_size, activation=tf.nn.softmax, name="output")

    return h2, logP

Within the session defined in the training method, the loss and softmax are determined as

_, gen_loss = sess.run((self.optimizer, self.generation_loss), feed_dict=feed)

and the softmax result is passed to the optimizer within the session as

self.opt.prob_dist = self.dist.eval(session=sess)

In this approach, I find that when I check the prob_dist attribute of the optimizer by printing the attribute within the _apply_dense method, the attribute shows up as a Nonetype object. I am therefore wondering what I can do to pass the softmax output to the optimizer, particularly as a numpy array, so that it can use the softmax output as parameter in its weight update protocol.
